File: gunhoflash/longest-increasing-subsequence.py
import math
from typing import List

class Solution:
  def updateMinWithLowerbound(self, num, nums: List[int]):
    left = 0
    right = len(nums) - 1

    while left < right:
      midIndex = (left + right) >> 1
      midValue = nums[midIndex]

      if midValue < num:
        left = midIndex + 1
      else:
        right = midIndex

    nums[left] = num

  def lengthOfLIS(self, nums: List[int]) -> int:
    lis = [math.inf]

    for num in nums:
      if num > lis[len(lis) - 1]:
        lis.append(num)
      else:
        self.updateMinWithLowerbound(num, lis)

    return len(lis)

File: gunhoflash/longest-increasing-subsequence.ts
function updateMinWithLowerbound(num: number, nums: number[]) {
  let left = 0, right = nums.length;

  while (left < right) {
    const midIndex = (left + right) >> 1;
    const midValue = nums[midIndex];

    if (midValue >= num) {
      right = midIndex;
    } else {
      left = midIndex + 1;
    }
  }

  nums[left] = num;
}

function lengthOfLIS(nums: number[]): number {
  const lis = [Infinity];
  nums.forEach(num => updateMinWithLowerbound(num, lis));
  return lis.length;
}
